Posts/2023/09/List-OS-Object-Names.md

```diff
@@ -0,0 +1,449 @@
+        // this function is defined, if we want to set the last error,
+        // and optionally the encoding.
+        [DllImport("ntdll.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern int NtQueryInformationProcess(
+            IntPtr ProcessHandle,
+            PROCESSINFOCLASS ProcessInformationClass,
+            IntPtr ProcessInformation,
+            uint ProcessInformationLength,
+            out uint ReturnLength
+        );
+
+        // This is similar to the previous one, but returns information about
+        // objects being used by processes.
+        [DllImport("ntdll.dll", SetLastError = true, CharSet = CharSet.Unicode)]
+        public static extern int NtQueryObject(
+            IntPtr Handle,
+            OBJECT_INFORMATION_CLASS ObjectInformationClass,
+            IntPtr ObjectInformation,
+            uint ObjectInformationLength,
+            out uint ReturnLength
+        );
+
+        // This function opens a handle to a process. We need to use it instead
+        // of the .NET APIs because the handles it uses are only references to
+        // native handles, and not the handles themselves.
+        [DllImport("kernel32.dll", SetLastError = true)]
+        public static extern IntPtr OpenProcess(
+             uint processAccess,
+             bool bInheritHandle,
+             uint processId
+        );
+
+        // This function duplicates a handle opened by a process.
+        // We need to duplicate the handle to be able to query it.
+        [DllImport("kernel32.dll", SetLastError=true)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        public static extern bool DuplicateHandle(
+            IntPtr hSourceProcessHandle,
+            IntPtr hSourceHandle,
+            IntPtr hTargetProcessHandle,
+            out IntPtr lpTargetHandle,
+            uint dwDesiredAccess,
+            [MarshalAs(UnmanagedType.Bool)] bool bInheritHandle,
+            uint dwOptions
+        );
+
+        // This function returns a handle to the current process.
+        // We need it so the OS can associate the duplicated handle
+        // with it.
+        [DllImport("kernel32.dll", SetLastError = true)]
+        public static extern IntPtr GetCurrentProcess();
+
+        // This function closes a handle opened by certain functions.
+        // Is important to close all handles where necessary to avoid
+        // memory leaks.
+        [DllImport("kernel32.dll", SetLastError = true)]
+        public static extern bool CloseHandle(IntPtr hObject);
+    }
+}
+```
+
+## The script
+
+### Adding the signature type
+
+The first thing we need to do is add that signature as a type in our PowerShell script. There are
+two ways we can do it, by copying that C# text into a here-string, or saving it in a file.
+The second one keeps the code cleaner, so for this post, let's assume I saved that file as
+`C:\HandleObjectNames.cs`.
```

Comment:
```suggestion
The first thing we need to do is add that signature as a type in our PowerShell script. Because we
defined the code in separate files, we can use the `Add-Type` cmdlet with the **Path** parameter to
compile them together and add the `Utilities` namespace.
```